mproton_darwin.m

```diff
@@ -0,0 +1,255 @@
+// +build darwin
+
+//  main.m
+//  cmini
+//
+//  Created by Igor Okulist on 5/27/21.
+//
+
+#import <Foundation/Foundation.h>
+#import <AppKit/AppKit.h>
+#import <WebKit/WebKit.h>
+
+#include "mproton.h"
+
+@interface AppContext : NSObject <WKScriptMessageHandlerWithReply>
+@property(nonatomic, retain)  NSWindow  * _Nullable window;
+@property(nonatomic, retain)  NSMenu  * _Nullable mainMenu;
+@property(nonatomic, retain)  NSStatusItem * _Nullable statusItem;
+@property(nonatomic, retain)  WKWebView * _Nullable webView;
+
+- (void)userContentController:(nonnull WKUserContentController *)userContentController
+      didReceiveScriptMessage:(nonnull WKScriptMessage *)message
+                 replyHandler:(nonnull void (^)(id _Nullable, NSString * _Nullable))replyHandler;
+@end
+
+@implementation AppContext
+- (void)userContentController:(nonnull WKUserContentController *)userContentController
+      didReceiveScriptMessage:(nonnull WKScriptMessage *)message
+                 replyHandler:(nonnull void (^)(id _Nullable, NSString * _Nullable))replyHandler {
+    
+    NSLog(@"[objc] WKScriptMessageHandlerWithReply callback called");
+    
+    extern const char * goCallbackDispatcher(const void * _Nonnull, const char * _Nonnull);
+    
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0), ^{
```

Comment:
I think this is dangerous. With `dispatch_get_global_queue`, it'll get executed on some thread created by GCD. This thread is unknown to Go and I think it's not blessed to call back into Go from external threads. I say that for two reasons.
1. This old thread says so: https://groups.google.com/g/golang-nuts/c/H9QbEVrzccc/m/36lCu2Q4S7oJ
2. Looking at Go source, they seem to rely on TLS variable that points to a `G` struct that has stack extents associated with the current goroutine.

To play it safe, I would call into Go on the main thread. This can be either via `dispatch_get_main_queue` or just on the thread this function is being invoked. However I could not find docs so far as to what thread invokes this `userContentController` method.

Reply:
all info seems to be really old, the only place i can find actually discussing it is [1] some actual code mentioned by rgo [2]

[1] https://stackoverflow.com/questions/4312894/c-callbacks-and-non-go-threads

[2] https://github.com/rogpeppe/rog-go/tree/master/exp/callback